docs/src/main/docs/health/02_health_in_k8s.adoc

```diff
@@ -0,0 +1,217 @@
+
+Probes can be implemented with three different mechanisms:
+
+. running a command inside a container
+. making an `HTTP` request against a container
+. opening a `TCP` socket against a container
+
+A microservice exposed to HTTP will typically implement both the liveness probe
+ and the readiness probe using HTTP requests.
+
+Probes can be configured with different parameters, see the most relevant
+ ones below:
+
+[cols="2,5",role="flex, sm7"]
+|=======
+| `initialDelaySeconds`
+| Number of seconds before probes are initiated
+
+| `periodSeconds`
+| How often to perform the probe. Default to 10 seconds. Minimum value is 1
+
+| `timeoutSeconds`
+| Number of seconds after which the probe times out. Defaults to 1 second.
+ Minimum value is 1
+
+| `failureThreshold`
+| Number of failed probes before giving up. Defaults to 3. Minimum value is 1.
+|=======
+
+=== Liveness probe
+
+The liveness probe is used to verify the container has become unresponsive,
+ for example it can detect deadlocks or analyze heap usage. When Kubernetes
+ gives up on a liveness probe, the corresponding pod is restarted.
+
+NOTE: A bad liveness probe can result in repeated restarts.
+
+We recommend the following:
+
+* Avoid checking dependencies in a liveness probe
+* Set `timeoutSeconds` to avoid excessive probe failures
+* Acknowledge startup times with `initDelaysSeconds`
+
+=== Readiness probe
+
+The readiness probe is used to avoid routing requests to the pod until it is
+ ready to accept traffic. When Kubernetes gives up on a readiness probe, the
+ pod is not restarted, it is removed from the service routing.
+
+NOTE: A bad readiness probe can result in a cascading effect where all pods are
+ removed from the service routing.
+
+We recommend the following:
+
+* Be conservative when checking shared dependencies
+* Be aggressive when checking local dependencies
+* Set `failureThreshold` according to `periodSeconds` in order to accomodate
+ temporary errors
+
+== Troubleshooting probes
+
+Failed probes are recorded as events associated with their corresponding pods.
+ The event message will only contain the status code information.
+
+[source,bash]
+.Get the events of a single pod
+----
+POD_NAME=$(kubectl get pod -l app=acme -o jsonpath='{.items[0].metadata.name}') # <1>
+kubectl get event --field-selector involvedObject.name=${POD_NAME} # <2>
+----
+<1> Get the effective pod name by filtering pods with the label `app=acme`
+<2> Filter the events for the pod
+
+TIP: Create log messages in your health check implementation when setting a
+ `DOWN` status, this will allow you to correlate the cause of a failed probe.
+
+== Example
+
+The example below illustrates the usages of the Helidon health API for an
+ application that implements health endpoints for both probes as well as a
+ Kubernetes service and deployment.
+The application code dissociates the health endpoints from the default routes
+ so that they are not exposed by the service.
```

Comment:
I think "they" refers to "the health endpoints". For clarity, consider making that explicit.